src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala

```diff
@@ -687,3 +688,134 @@ private class BlockMatrixMultiplyRDD(l: BlockMatrix, r: BlockMatrix)
 }
 
 case class IntPartition(index: Int) extends Partition
+
+object WriteBlocksRDD {
+  // on return, deps(blockRowIndex) is the increasing array of (non-empty) parent partitions that overlap blockRow
+  def computeBlockRowDependencies(parentPartitionBoundaries: Array[Long], gp: GridPartitioner): Array[Array[Int]] = {
+    val rows = parentPartitionBoundaries.last
+    assert(rows == gp.rows)
+    val nBlockRows = gp.rowPartitions
+    val deps = Array.ofDim[Array[Int]](nBlockRows)
+    val ab = new ArrayBuilder[Int]()
+    
+    var i = 0 // parent partition index
+    var firstRowInBlock = 0L
+    var lastRowInBlock = 0L
+    var blockRowIndex = 0
+    while (blockRowIndex < nBlockRows) {
+      firstRowInBlock = lastRowInBlock
+      if (blockRowIndex != nBlockRows - 1)
+        lastRowInBlock += gp.blockSize
+      else
+        lastRowInBlock = parentPartitionBoundaries.last
+
+      // find all (non-empty) parent partitions overlapping this block
+      ab.clear()
+      while (parentPartitionBoundaries(i) < lastRowInBlock) {
+        if (parentPartitionBoundaries(i + 1) > firstRowInBlock)
+          ab += i
+        i += 1
+      }
+      // if last parent partition overlaps next blockRow, don't advance
+      if (parentPartitionBoundaries(i) > lastRowInBlock)
+        i -= 1
+      
+      deps(blockRowIndex) = ab.result()
+      blockRowIndex += 1
+    }
+
+    deps
+  }
+}
+
+class WriteBlocksRDD(irm: IndexedRowMatrix, path: String, blockSize: Int) extends RDD[Int](irm.rows.sparkContext, Nil) {
+  private val rows = irm.numRows()
+  private val cols = irm.numCols()
+  private val parentPartitions = irm.rows.partitions
+  private val parentPartitionBoundaries: Array[Long] = irm.rows.countPerPartition().scanLeft(0L)(_ + _)
+  
+  // all IndexedRows must be present in RDD
+  assert(rows == parentPartitionBoundaries.last,
+    s"IndexedRowMatrix has $rows rows but RDD only has ${parentPartitionBoundaries.last} IndexedRows.")
+
+  private val gp = GridPartitioner(blockSize, rows, cols)
+  private val blockRowDependencies = WriteBlocksRDD.computeBlockRowDependencies(parentPartitionBoundaries, gp)
+  private val truncatedBlockRow = rows / blockSize
+  private val truncatedBlockCol = cols / blockSize
+  private val excessRows = (rows % blockSize).toInt
+  private val excessCols = (cols % blockSize).toInt
+  
+  private val d = digitsNeeded(gp.numPartitions)
+  private val sHadoopBc = irm.rows.sparkContext.broadcast(
+    new SerializableHadoopConfiguration(irm.rows.sparkContext.hadoopConfiguration))
+
+  override def getDependencies: Seq[Dependency[_]] = {  
+    Array[Dependency[_]](
+      new NarrowDependency(irm.rows) {
+        def getParents(partitionId: Int): Seq[Int] = blockRowDependencies(partitionId)
+      }
+    )
+  }
+  
+  protected def getPartitions: Array[Partition] =
+    (0 until gp.rowPartitions).map(IntPartition).toArray[Partition]
```

Comment:
The "traditional" approach is to have a `WriteBlocksRDDPartition` that has the relevant information necessary to compute the partition, and then use that in `compute` and `getDependencies`.  See `BlockedRDD`, for example.  That avoids having to call `computeBlockRowDependencies` multiple times.